### DISCUSSION TITLE
window corruption/artifacts on long-running external command

### DISCUSSION THREAD

**Discussion by ITwrx** [can-lehmann/owlkettle#66] [Q&A]
I'm calling `rsync` after button click via `execCmdEx` from `std/osproc`. While these rsync processes are running, if i change focus to another desktop workspace (currently using xfce, but have seen this with gnome in the past) and go back to my owlkettle app, the window is corrupted with a black/dark grey color, or random artifacts until the rsync process(es) finish. Then, the gui returns to normal.

Is this an owlkettle/gtk4/x11/wayland issue? or i'm supposed to use spawn, or async, or something for the rsync process(es) to avoid this issue? 

thanks

**Comment by ITwrx** [comment]
yeah, i think it's number two. It also explains another issue i'm having. I guess i thought there would be some hidden magic for the lazy or something... but execCmdEx is synchronous, so it's blocking my gui. I will leave this open, for now, in case anyone has any recommendations as to the best approach. Actually, rsync is usually getting called multiple times inside a loop, so maybe it's a chance for some performance optimization too.

**Reply by can-lehmann** [reply]
Yes, blocking inside an event handler will break your UI. As far as I can tell, [`startProcess`](https://nim-lang.org/docs/osproc.html#startProcess%2Cstring%2Cstring%2CopenArray%5Bstring%5D%2CStringTableRef%2Cset%5BProcessOption%5D) is non-blocking, and should be a viable alternative.

**Reply by ITwrx** [reply]
Thanks for the response. I didn't know how to use `startProcess` and still get the output and exit code from `rsync` like the `execCmdEx` conveniently provides. No `async execCmdEx` and [this async proc](https://github.com/cheatfate/asynctools/blob/master/asynctools/asyncproc.nim#L122) didn't work for me. Looks abandoned. That's OK, i just implemented threading, largely thanks to the owlkettle threading example.
